File: Module 1/Renderer.cpp
#include "Renderer.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

#ifdef _WIN32
#include <cstdlib>
#endif

Renderer::Renderer(int widthVal, int heightVal) : width(widthVal), height(heightVal) {}

char Renderer::glyphFor(double mass, double minLogMass, double maxLogMass) {
    static const std::string glyphs = ".+*O@";
    if (mass <= 0.0 || std::abs(maxLogMass - minLogMass) < 1e-12) {
        return '*';
    }

    const double ratio = (std::log10(mass) - minLogMass) / (maxLogMass - minLogMass);
    int idx = static_cast<int>(ratio * (glyphs.size() - 1) + 0.5);
    idx = std::clamp(idx, 0, static_cast<int>(glyphs.size() - 1));
    return glyphs[static_cast<std::size_t>(idx)];
}

void Renderer::render(
    const std::vector<Body>& bodies,
    double timeElapsed,
    double energy,
    const std::vector<std::string>& collisionLog) const {
#ifdef _WIN32
    std::system("cls");
#else
    std::cout << "\x1B[2J\x1B[H";
#endif

    if (bodies.empty()) {
        std::cout << "(no bodies to render)\n";
        return;
    }

    double minX = bodies[0].position.x;
    double maxX = bodies[0].position.x;
    double minY = bodies[0].position.y;
    double maxY = bodies[0].position.y;

    std::vector<double> logMasses;
    logMasses.reserve(bodies.size());

    for (const auto& body : bodies) {
        minX = std::min(minX, body.position.x);
        maxX = std::max(maxX, body.position.x);
        minY = std::min(minY, body.position.y);
        maxY = std::max(maxY, body.position.y);
        if (body.mass > 0.0) {
            logMasses.push_back(std::log10(body.mass));
        }
    }

    double spanX = std::max(maxX - minX, 1.0);
    double spanY = std::max(maxY - minY, 1.0);

    minX -= spanX * 0.12;
    maxX += spanX * 0.12;
    minY -= spanY * 0.12;
    maxY += spanY * 0.12;

    spanX = maxX - minX;
    spanY = maxY - minY;

    const double desiredRatio = (static_cast<double>(width) / static_cast<double>(height)) * 0.5;
    if ((spanX / spanY) < desiredRatio) {
        const double cx = (minX + maxX) * 0.5;
        spanX = spanY * desiredRatio;
        minX = cx - spanX * 0.5;
        maxX = cx + spanX * 0.5;
    } else {
        const double cy = (minY + maxY) * 0.5;
        spanY = spanX / desiredRatio;
        minY = cy - spanY * 0.5;
        maxY = cy + spanY * 0.5;
    }

    const double minLogMass = *std::min_element(logMasses.begin(), logMasses.end());
    const double maxLogMass = *std::max_element(logMasses.begin(), logMasses.end());

    std::vector<std::string> grid(static_cast<std::size_t>(height), std::string(static_cast<std::size_t>(width), ' '));

    for (const auto& body : bodies) {
        int col = static_cast<int>(((body.position.x - minX) / (maxX - minX)) * (width - 1));
        int row = static_cast<int>((1.0 - (body.position.y - minY) / (maxY - minY)) * (height - 1));

        col = std::clamp(col, 0, width - 1);
        row = std::clamp(row, 0, height - 1);

        grid[static_cast<std::size_t>(row)][static_cast<std::size_t>(col)] = glyphFor(body.mass, minLogMass, maxLogMass);
    }

    std::cout << '+' << std::string(static_cast<std::size_t>(width), '-') << "+\n";
    for (const auto& row : grid) {
        std::cout << '|' << row << "|\n";
    }
    std::cout << '+' << std::string(static_cast<std::size_t>(width), '-') << "+\n";

    const double days = timeElapsed / 86400.0;
    const double years = days / 365.25;

    std::cout.setf(std::ios::scientific);
    std::cout.precision(4);

    if (years >= 1.0) {
        std::cout.unsetf(std::ios::scientific);
        std::cout << " Time: " << years << " yr";
    } else if (days >= 1.0) {
        std::cout.unsetf(std::ios::scientific);
        std::cout << " Time: " << days << " d";
    } else {
        std::cout.unsetf(std::ios::scientific);
        std::cout << " Time: " << timeElapsed << " s";
    }

    std::cout.setf(std::ios::scientific);
    std::cout << "  |  Bodies: " << bodies.size() << "  |  Total E: " << energy << " J\n";

    if (!collisionLog.empty()) {
        std::cout << " Last event: " << collisionLog.back() << "\n";
    }
    std::cout << " Legend: '.' small '+' medium '*' large 'O' giant '@' massive\n";
}

File: Module 1/renderer.py
"""
renderer.py — ASCII terminal renderer for the N-body simulation.

Days 8-9: maps body positions onto a character grid, clears the terminal
          and prints one frame per simulation snapshot.
"""

import os
import math
from body import Body


# Glyph scale by relative mass (log scale): tiny → massive
_GLYPHS = [".", "+", "*", "O", "@"]


def _glyph_for(body: Body, max_log_mass: float, min_log_mass: float) -> str:
    """Pick a display character based on the body's relative log-mass."""
    span = max_log_mass - min_log_mass
    if span == 0 or body.mass <= 0:
        return "*"
    ratio = (math.log10(body.mass) - min_log_mass) / span
    idx = int(ratio * (len(_GLYPHS) - 1) + 0.5)
    idx = max(0, min(idx, len(_GLYPHS) - 1))
    return _GLYPHS[idx]


class Renderer:
    """
    Renders one simulation frame on a fixed-size ASCII grid.

    The viewport auto-scales every frame so all bodies stay visible.
    Terminal character cells are roughly twice as tall as wide, so the
    grid is kept at a 2:1 width-to-height ratio to appear square.
    """

    def __init__(self, width: int = 78, height: int = 36):
        self.width = width
        self.height = height

    # ------------------------------------------------------------------ #
    # Viewport helpers                                                      #
    # ------------------------------------------------------------------ #

    def _bounds(
        self, bodies: list[Body]
    ) -> tuple[float, float, float, float]:
        """Return (min_x, max_x, min_y, max_y) with a 10 % margin, kept square."""
        xs = [b.position.x for b in bodies]
        ys = [b.position.y for b in bodies]
        min_x, max_x = min(xs), max(xs)
        min_y, max_y = min(ys), max(ys)

        span_x = max(max_x - min_x, 1.0)
        span_y = max(max_y - min_y, 1.0)

        min_x -= span_x * 0.12
        max_x += span_x * 0.12
        min_y -= span_y * 0.12
        max_y += span_y * 0.12

        span_x = max_x - min_x
        span_y = max_y - min_y

        # Correct for the ~2:1 char aspect ratio
        desired_ratio = (self.width / self.height) * 0.5
        if span_x / span_y < desired_ratio:
            cx = (min_x + max_x) / 2
            span_x = span_y * desired_ratio
            min_x, max_x = cx - span_x / 2, cx + span_x / 2
        else:
            cy = (min_y + max_y) / 2
            span_y = span_x / desired_ratio
            min_y, max_y = cy - span_y / 2, cy + span_y / 2

        return min_x, max_x, min_y, max_y

    # ------------------------------------------------------------------ #
    # Public render                                                         #
    # ------------------------------------------------------------------ #

    def render(
        self,
        bodies: list[Body],
        time_elapsed: float,
        energy: float | None = None,
        collision_log: list[str] | None = None,
    ) -> None:
        """Clear the terminal and draw one frame."""
        os.system("cls" if os.name == "nt" else "clear")

        if not bodies:
            print("(no bodies to render)")
            return

        min_x, max_x, min_y, max_y = self._bounds(bodies)
        span_x = max_x - min_x
        span_y = max_y - min_y

        # Pre-compute glyph scaling
        log_masses = [math.log10(b.mass) for b in bodies if b.mass > 0]
        max_lm = max(log_masses)
        min_lm = min(log_masses)

        # Build empty grid (background = space)
        grid: list[list[str]] = [[" " for _ in range(self.width)] for _ in range(self.height)]

        # Plot bodies — later bodies overwrite earlier if they overlap
        for body in bodies:
            col = int((body.position.x - min_x) / span_x * (self.width - 1))
            row = int((1.0 - (body.position.y - min_y) / span_y) * (self.height - 1))
            col = max(0, min(col, self.width - 1))
            row = max(0, min(row, self.height - 1))
            grid[row][col] = _glyph_for(body, max_lm, min_lm)

        # Print with border
        border = "+" + "-" * self.width + "+"
        print(border)
        for row in grid:
            print("|" + "".join(row) + "|")
        print(border)

        # Status bar
        days = time_elapsed / 86400.0
        years = days / 365.25
        if years >= 1.0:
            time_str = f"{years:.2f} yr"
        elif days >= 1.0:
            time_str = f"{days:.1f} d"
        else:
            time_str = f"{time_elapsed:.1f} s"

        status = f" Time: {time_str}  |  Bodies: {len(bodies)}"
        if energy is not None:
            status += f"  |  Total E: {energy:.4e} J"
        print(status)

        # Show last collision if any
        if collision_log:
            print(f" Last event: {collision_log[-1]}")

        # Legend
        print(f" Legend: '.' small  '+' medium  '*' large  'O' giant  '@' massive")
